src/core/ui/mainui.cpp

```diff
@@ -14,36 +16,76 @@ namespace otto::ui {
 
     } empty_screen;
 
+    std::map<std::string, std::function<engines::AnyEngine *()>> engines = {
+      { "TapeDeck", []() { return (engines::AnyEngine*)&global::tapedeck; } },
+      { "Mixer", []() { return (engines::AnyEngine*)&global::mixer; } },
+      { "Synth", []() { return (engines::AnyEngine*)&*global::synth; } },
+      { "Drums", []() { return (engines::AnyEngine*)&*global::drums; } },
+      { "Metronome", []() { return (engines::AnyEngine*)&global::metronome; } }
+    };
+
+    std::string selected_engine_name = "";
     Screen* cur_screen = &empty_screen;
 
     PressedKeys keys;
-
   }
 
   bool is_pressed(Key k) noexcept
   {
     return keys[static_cast<unsigned>(k)];
   }
 
+  void selectEngine(std::string engine_name) {
+    selected_engine_name = engine_name;
+
+    auto getter = engines[engine_name];
+    if (!getter) {
+      getter = engines[INITIAL_ENGINE];
+      selected_engine_name = INITIAL_ENGINE;
+    }
+
+    auto engine = getter();
+
+    display(engine->screen());
+  }
+
+  void init() {
+    selectEngine(selected_engine_name);
+  }
+
   void display(Screen& screen)
   {
     cur_screen->on_hide();
     cur_screen = &screen;
     cur_screen->on_show();
   }
 
+  nlohmann::json serialize() {
```

Comment:
these should be called `to_json` and `from_json`, my mistake

Reply:
Done